### PR TITLE
Fix #552 #560

### LINKED ISSUE #552: distance2boundary: Result is not checked for NaN
During the calculation of the distance to boundary it is assumed that the packet's radius is between r_inner and r_outer. Because of rounding errors this is not always true and there exists a small range of parameters (radius and mu) for which the discriminant gets negative (the packed doesn't hit the boundary). In that case the distance is NaN.

Possible ways to fix this:
- When crossing the boundary explicitly set the radius of the packet
- If dboundary is NaN, set it to a constant maximum (2 \* r_outer for example)
- Return some TARDIS_ERROR and discard packet

### PR DESCRIPTION
**Pull request by yeganer**
This aims to fix Issue #552 and #560.

The code from move_packet is mirrored in move_packet_across_boundary.  But instead of calculating the new position it's known value is used.
The current implementation relies on rpacket.d_boundary being set only for the calculation of the estimators.

### CODE REVIEW

File: tardis/montecarlo/src/cmontecarlo.c [hunk 1]
```diff
@@ -234,6 +248,7 @@ compute_distance2line (rpacket_t * packet, const storage_model_t * storage)
           fprintf (stderr, "nu = %f\n", nu);
           fprintf (stderr, "doppler_factor = %f\n", doppler_factor);
           fprintf (stderr, "cur_zone_id = %" PRIi64 "\n", cur_zone_id);
+          rpacket_set_d_line (packet, NAN);
```
Comment (unoebauer) [inline tardis/montecarlo/src/cmontecarlo.c]: Is there a reason why we set d_line to NAN and not any or all of the other distances?

File: tardis/montecarlo/src/cmontecarlo.c [hunk 2]
```diff
@@ -489,7 +504,37 @@ void
...
+          comov_energy * distance;
+#ifdef WITHOPENMP
+#pragma omp atomic
+#endif
+      storage->nubars[rpacket_get_current_shell_id (packet)] +=
+          comov_energy * distance * comov_nu;
+    }
+  // done moving packet
+
+  // reset tau event
```
Comment (unoebauer) [inline tardis/montecarlo/src/cmontecarlo.c]: Why is this necessary, why not keep move_packet here? **DRY**

File: tardis/montecarlo/src/cmontecarlo.c (hunk 1 above)
Comment (yeganer) [inline tardis/montecarlo/src/cmontecarlo.c]: My reasoning was, that since the calculation of `distance2line` fails we set the corresponding distance to NaN. Setting `d_boundary` in a function that calculates `d_line` seems counter intuitive to me.

The proper 'solution' would be to check the error of the distance calculations and terminate if something goes wrong.

**Comment by unoebauer** [comment]
Quite a few changes for a simple fix, but fine. However, I have a few remarks (see inline comments). In particular, I am not happy about copying the contents of move_packet into the cross_boundary function (Don't repeat yourself).

I'd suggest one of the following:
* leave move_packet in there and just set distance explicitly afterwards; this will introduce a slight inconsistency between the distance to boundary used in the estimator increments and the position of the packet; however, **if** the deviations between the shell interface location and the packet location are on the level of machine precision, these discrepancies should be of the same order
* modularize the move packet routine - separate actual packet propagation from estimator increments, distance recalculation, etc

I'd favour the latter approach...

**Comment by unoebauer** [comment]
Following up on the thoughts from above, I wonder whether it may be prudent to introduce a sanity check here. In particular ensuring that the relative difference between the packet location determined traditionally (move packet) and the interface location is not larger than, say 1e-10. This way, we would make sure that we only do this explicit boundary fix for cases in which machine precision introduces discrepancies and not in cases where some other error introduces differences. Thoughts @ssim, @wkerzendorf, @chvogl, @yeganer?

File: tardis/montecarlo/src/cmontecarlo.c (hunk 2 above)
Comment (yeganer) [inline tardis/montecarlo/src/cmontecarlo.c]: I asked myself the same question and came to the conclusion that `move_packet` and `move_packet_across_boundary` are two different functions which happen to have some bits in common.

- `move_packet` just moves the packet a known distance but 
- `move_packet_across_boundary` moves the packet TO a known point.
To me that's two different things. In theory the right thing to do would be to calculate `distance2boundary` again but we know that `d_boundary` didn't change between the last calculation and calling `move_packet_across_boundary` so I left it out.
Moving the packet and then setting the radius explicitly would work of course but that would repeat the calculation while using existing code. So both variants have their trade-offs.

The best solution in my opinion would be to have a function `increment_estimators` which is called by both methods. This would keep the (different) logic separated but combine the common one.

**Comment by yeganer** [comment]
@unoebauer 
I think a sanity check isn't necessary because we use the exact same formula to calculate the `distance2boundary` and use the result to calculate the new_r. Mathematically we can prove that `new_r == r_boundary`. Because of machine precision this is not the case. But if we assume similiar orders of magnitude for `r` and `r_boundary` we are only doing errors on the level of machine precision.

**Comment by unoebauer** [comment]
@yeganer  - I am aware of that. However, my concern was more that due some possible future changes we may create situations in which we introduce differences between the calculated r_packet and the expected one. This discrepancy would be absorbed by our quickfix (which was not designed to compensate this type of error). However, if we all think that the unit tests which are checking all c Monte Carlo functions would catch such a situation, we can also not explicitly include an additional sanity check.

File: tardis/montecarlo/tests/test_cmontecarlo.py [hunk 3]
```diff
@@ -65,6 +66,8 @@
 cmontecarlo_filepath = os.path.join(path[0], 'montecarlo', 'montecarlo.so')
 cmontecarlo_methods = CDLL(cmontecarlo_filepath)
 
+INVERSE_C = 3.33564095198152e-11
```
Comment (wkerzendorf) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: I think we should get those from astropy at some stage.

Comment (yeganer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: Sure but when I did that the tests failed so I hardcoded the value from the C part

Comment (unoebauer) [inline tardis/montecarlo/tests/test_cmontecarlo.py]: At least add a comment detailing the current need to hardwire this.

**Comment by yeganer** [comment]
@tardis-sn/tardis-core This PR is ready to be merged from my side.

**Comment by wkerzendorf** [comment]
We are not merging this - right? For the moment, we will only use this as a branch to do dalek at the moment?

**Comment by yeganer** [comment]
@wkerzendorf I'm not in a position to decide whether to merge or not. Although the underlying bug is affecting tardis and could happen during any run I'm completely fine if we postpone merging and I'm running dalek on this branch instead of master.

**Comment by wkerzendorf** [comment]
merge when integration tests.

**Comment by unoebauer** [comment]
Since the integration tests are up and running, we should merge this. @yeganer can you resolve the conflicts?
